qa/L0_response_cache/ensemble_cache_test.py

```diff
@@ -0,0 +1,319 @@
+        self._check_valid_stats(model_inference_stats)
+        self._check_single_cache_miss_success_inference(model_inference_stats)
+        cached_outputs = self.inference_handler.infer()
+        self._check_valid_output(cached_outputs)
+        self._check_valid_output_inference(inference_outputs, cached_outputs)
+        model_inference_stats = self.inference_handler.get_inference_statistics(model)
+        self._check_valid_stats(model_inference_stats)
+        return model_inference_stats
+
+
+class EnsembleCacheTest(tu.TestResultCollector):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.config_manager = ModelConfigManager()
+        self.inference_validator = InferenceValidator()
+        self.inference_handler = InferenceHandler()
+        self.ensemble_model = self.config_manager.get_ensemble_model()
+        self.composing_model = self.config_manager.get_composing_model()
+
+    def setup_cache_ensemble_model(self):
+        self.config_manager.enable_response_cache_ensemble_model()
+
+    def setup_decoupled_ensemble_model(self):
+        self.config_manager.enable_decoupled_ensemble_model()
+
+    def setup_decoupled_composing_model(self):
+        self.config_manager.enable_decoupled_composing_model()
+
+    def setup_cache_composing_model(self):
+        self.config_manager.enable_response_cache_composing_model()
+
+    def setup_cache_all_models(self):
+        self.config_manager.enable_response_cache_all_models()
+
+    def reset_config(self):
+        self.config_manager.reset_config_files()
+
+    def test_ensemble_top_level_cache(self):
+        model_inference_stats = self.inference_validator.run_ensemble(
+            self.ensemble_model
+        )
+        self.assertNotEqual(
+            model_inference_stats["cache_hit"]["count"],
+            "0",
+            "Unexpected error in top-level ensemble response caching",
+        )
+        self.assertLessEqual(
+            int(model_inference_stats["cache_hit"]["count"]),
+            1,
+            "Unexpected multiple cache hits",
+        )
+        self.reset_config()
+
+    def test_all_models_with_cache_enabled(self):
+        ensemble_model_stats = self.inference_validator.run_ensemble(
+            self.ensemble_model
+        )
+        composing_model_stats = self.inference_handler.get_inference_statistics(
+            self.composing_model
+        )
+        self.assertNotEqual(
+            ensemble_model_stats["cache_hit"]["count"],
+            "0",
+            "Unexpected error in top-level ensemble request caching",
+        )
+        self.assertNotIn(
+            "count",
+            composing_model_stats["cache_hit"],
+            "Unexpected composing model cache hits",
+        )
+        self.reset_config()
+
+    def test_composing_model_cache_enabled(self):
+        composing_model_stats = self.inference_validator.run_ensemble(
+            self.composing_model
+        )
```

Comment:
is `composing_model` also an ensemble?

Reply:
Actually, the run_ensemble runs the ensemble model but the self.composing_model parameter is to define which model stats to verify. I'm doing a bunch of stats verification in run_ensemble. In the case of response_cache enable only in composing model, the top_level ensemble model stats will not have any cache statistics. 
I can see that it's cofusing. I will modify the function name to avoid confusion.